app/services/ticker.py

```diff
@@ -0,0 +1,183 @@
+                    "bid" = $5,
+                    "ask" = $6,
+                    "updated" = $7
+                WHERE tickers.token_address = $1
+            """,
+            Web3.toBytes(hexstr=ticker_info["token_address"]),
+            ticker_info["quote_volume"],
+            ticker_info["base_volume"],
+            ticker_info["last"],
+            ticker_info["bid"],
+            ticker_info["ask"],
+            datetime.utcnow())
+
+async def update_ticker(token_addr):
+    current_block = App().web3.eth.blockNumber
+    coin_contract = ERC20Token(token_addr)
+    base_contract = ERC20Token(ZERO_ADDR)
+
+    ticker_info = { "token_address": token_addr }
+
+    volumes = await get_trades_volume(token_addr)
+    ticker_info.update({
+        "quote_volume": coin_contract.denormalize_value(volumes["quote_volume"]),
+        "base_volume": base_contract.denormalize_value(volumes["base_volume"]),
+    })
+
+    trade = await get_last_trade(token_addr)
+    if trade:
+        side = "buy" if trade["token_get"] == ZERO_ADDR_BYTES else "sell"
+        # Compute the price.
+        try:
+            if side == "buy":
+                price = base_contract.denormalize_value(trade["amount_get"]) \
+                    / coin_contract.denormalize_value(trade["amount_give"])
+            else:
+                price = base_contract.denormalize_value(trade["amount_give"]) \
+                    / coin_contract.denormalize_value(trade["amount_get"])
+        except (InvalidOperation, DivisionByZero):
+            # Somewhere, somehow, with all our high-precision math, we still get
+            # rounding. Or maybe there are somehow 0-trades.
+            logger.debug("Failed to compute price: token=%s side=%s txid=%s logidx=%i amount_get=%s amount_give=%s",
+                token_addr, side, Web3.toHex(trade["transaction_hash"]), trade["log_index"], trade["amount_get"], trade["amount_give"])
+            ticker_info["last"] = None
+        else:
+            ticker_info["last"] = price
+    else: # No last price available
+        ticker_info["last"] = None
+
+    # TODO: percent_change
+
+    spread = await get_market_spread(token_addr, current_block)
+    if spread["ask"]:
+        # Spread comes back in token/base native values, denormalize it here
+        ticker_info["ask"] = spread["ask"] * base_contract.denormalize_value(1.0) \
+                                / coin_contract.denormalize_value(1.0)
+    else:
+        ticker_info["ask"] = None
+
+    if spread["bid"]:
+        ticker_info["bid"] = spread["bid"] * base_contract.denormalize_value(1.0) \
+                                / coin_contract.denormalize_value(1.0)
+    else:
+        ticker_info["bid"] = None
+
+    logger.debug(ticker_info)
+    await save_ticker(ticker_info)
+
+
+async def main():
+    while True:
+        token = tokens_queue.get()
+        await update_ticker(token)
+        tokens_queue.put(token)
+        await asyncio.sleep(2.0)
```

Comment:
How did you arrive at 2 seconds?

Reply:
"Update every ticker about every half hour", with N_tokens = 1000 gives us ~2 seconds, ignoring processing time. Open for updates and discussion though.

Reply:
Can we run multiple threads of this?

Reply:
We could. Do we need to?